src/test/java/org/owasp/esapi/reference/validation/BaseValidationRuleTest.java

```diff
@@ -15,87 +15,274 @@
+
+    @Test
+    public void testIsValidOnExceptionRetursFalse() throws ValidationException {
+        when(uit.getValid(STR_VAL, STR_VAL)).thenThrow(testValidationEx);
+        assertFalse(uit.isValid(STR_VAL, STR_VAL));
+        verify(uit, times(1)).getValid(STR_VAL, STR_VAL);
+    }
+
+    /* *************************
+     * TO DISCUSS 
+     * FIXME
+     * Tests below this block are items which are valid against the current implementation, but have side effects
+     * or unclear results under certain conditions.
+     * 
+     * Once Items are discussed and understood they should probably be well-commented and moved out of this area.
+     */
+
+    @Test
+    public void testGetValidMultipleExceptionSameContextThrowsRuntimeException() throws ValidationException {
+        exEx.expect(RuntimeException.class);
+        ValidationErrorList vel = new ValidationErrorList();
+        when(uit.getValid(STR_VAL, STR_VAL)).thenThrow(testValidationEx);
+        uit.getValid(STR_VAL, STR_VAL, vel);
+        /*
+         * Side-effect of ValidationErrorList. If the same context is used against a BaseValidationRule multiple times resulting in exception, the ValidationErrorList impl will blow up.
+         * This can be an unclear event at runtime if a single BaseValidationRule instance is shared in an application and multiple parts happen to use the same contextual string to capture failure events.
+         * 
+         */
+        uit.getValid(STR_VAL, STR_VAL, vel);
+    }
+
+    //None of the Whitelist content belongs in this class, IMO.
+    
+    @Test
+    public void testWhitelistCharArrayCleansString() {
+        String myString = "AAAGaaadBBB12345*";
+        char[] whitelist = new char[] {'d', 'G', '3', '*', ']'};
+        String result = uit.whitelist(myString, whitelist);
+        assertEquals("Gd3*", result);
+    }
+
+    @Test
+    public void testWhitelistNullCharArrayThrows() {
+        exEx.expect(NullPointerException.class);
+        String myString = "AAAGaaadBBB12345*";
+        char[] whitelist = null;
+        uit.whitelist(myString, whitelist);
+    }
+
+    @Test
+    public void testWhitelistSetCleansString() {
+        String myString = "AAAGaaadBBB12345*";
+        Set<Character> whitelist = new HashSet<>();
+        whitelist.add('d');
+        whitelist.add('G');
+        whitelist.add('3');
+        whitelist.add('*');
+        whitelist.add(']');
+        String result = uit.whitelist(myString, whitelist);
+        assertEquals("Gd3*", result);
+    }
+
+    @Test
+    public void testWhitelistNullSetThrows() {
+        exEx.expect(NullPointerException.class);
+        String myString = "AAAGaaadBBB12345*";
+        Set<Character> whitelist = null;
+        uit.whitelist(myString, whitelist);
+    }
+
+    @Test
+    public void testWhitelistSetExtendedCharacterSets() {
+        String myString = "𡘾𦴩<𥻂";
+        //This was an issue wiht HTMLEntityCodec at one point, so it seems like we should be doing this test.
```

Comment:
@xeno6696 Since you worked on the HTMLEntityCodec I thought you may have context into this test.  
This test is currently marked as Ignored, but it feels like something that should be verified.  
Interested in your thoughts.

Reply:
I see you got the string from the HTMLEntityCodecTest...  I'm not sure why you didn't copy over the actual test....

Reply:
```
	@Test
	public void testEntityDecoding(){
		assertEquals("<", codec.decode("&lt;"));
        assertEquals( "<", codec.decode("&LT"));
        assertEquals( "<", codec.decode("&lt;"));
        assertEquals( "<", codec.decode("&LT;"));
	}
	
	@Test
	public void test32BitCJK(){
		String s = "𡘾𦴩𥻂";
		String expected = "&#x2163e;&#x26d29;&#x25ec2;";
		String bad = "&#xd845;&#xde3e;&#xd85b;&#xdd29;&#xd857;&#xdec2;";
		assertEquals(false, expected.equals(bad));
		assertEquals(expected, codec.encode(new char[0], s));
	}
	
	@Test
	public void test32BitCJKMixedWithBmp(){
		String s = "𡘾𦴩<𥻂";
		String expected = "&#x2163e;&#x26d29;&lt;&#x25ec2;";
		String bad = "&#xd845;&#xde3e;&#xd85b;&#xdd29;&#xd857;&#xdec2;";
		assertEquals(false, expected.equals(bad));
		assertEquals(expected, codec.encode(new char[0], s));
	}
	
	@Test
	public void testDecodeforChars(){
		String s = "!@$%()=+{}[]";
		String expected = "!@$%()=+{}[]";
		assertEquals(expected, codec.decode(s));
	}
	
	@Test
	public void testMixedBmpAndNonBmp(){
		String nonBMP = new String(new int[]{0x2f804}, 0, 1);
		String bmp = "<a";
		String expected = "&lt;a&#x2f804;";
		String input = bmp + nonBMP;
		assertEquals(expected, codec.encode(new char[0], input));
	}
```

Reply:
obviously that very first test has a couple of mistakes that we missed, but these are the core tests.

Reply:
In this case we're not encoding/decoding.  We're stripping characters from a string. 
I believe I have reasonable tests for the UTF-8 characters, but the utf-16 gave me some heartburn.

I think I tried to remove one of the characters from the origin string and ended up creating a test failure.  Since this was ancillary to the fix I was chasing I used the fail clause to point out that I wasn't sure what exactly I was doing, or trying to do,  so I could follow up on it later.

I think I was writing a bad test, or I had bad expectations with the character whitelist behavior.  I can take another crack at it.  Looking at it again, I think I need to whitelist the character pair that make up the symbol display.  That may suffice for showing that utf-16 can be whitelisted through BasicValidationRule.

I don't think the other tests add value here, which is why I only pulled this data set.

Reply:
Yeah, I probably was not treating the utf-16 correctly.  This implementation shows my desired intention:

 ```
@Test
    
    public void testWhitelistSetExtendedCharacterSets() {
        String myString = "𡘾𦴩<𥻂";
        //(55365 56894) (55387 56617) 60 (55383 57026)
        Set<Character> whitelist = new HashSet<>();
        whitelist.add((char) 60);
        whitelist.add((char) 55387);
        whitelist.add((char) 56617);
        String result = uit.whitelist(myString, whitelist);
        assertEquals("𦴩<", result);
    }
```

Reply:
This test is added to the pull request.